### PR TITLE
Further lute code (mgavioli / adjusted by MarkusLutz63)

### PR DESCRIPTION
**Pull request by MarkusLutz63**
With this PR  (and the PR before: #3064) lute code is more or less on the status of dev!

1. TAB - Mixing mensural value symbols and beaming in historic tablatures
2. TAB - More French styles for frets and duration symbols
3. TAB - Font design improvement
4. TAB: New historical English fret style

TAB - Mixing mensural value symbols and beaming in historic tablatures

__References__: Technology Preview forum post with discussion, screen-shots and links to additional threads https://musescore.org/en/node/81051

Implements the possibility to mix, in TAB's with note symbols at staff side, mensural value symbols (discreet glyphs) with the 'grid'-shaped beaming found in historical sources and commonly used in lute literature.

The beaming is implemented by special drawing of the `TabDurationSymbol` element holding the note value symbol, using drawing primitives instead of the relevant font glyph.

The user may choose between the two renderings (discreet glyph or beaming grid), on a chord-by-chord basis, by setting the chord `BeamMode` (via for instance the relevant palette): `AUTO` selects the glyph rendering, `beam start` the start of the grid and `beam middle` the continuation of the grid beamed to the previous element.

Typographic features of the grid (stem width, stem height and beam thickness) depend on the glyph style and are hard-coded in the style definition. As well as the number of beams for note value, which also depends on the note value style.

Also:
- Implements to possibility to force the display of a note value which would not be rendered by the current note value repetition setting, by setting the chord beam mode to any other value.
- Implements the 'no stem' chord setting for this TAB style, allowing to remove a note value symbol otherwise generated by the current repetition setting.
- Improves the detection of note value font metrics (still not perfect, though, as Qt `QFontMetricsF::tightboundingRect()` returns very approximated results)
- Fixes note value glyph scaling, when the staff scale is modified.

2. TAB - More French styles for frets and duration symbols

__Reference__: discussion, screen-shots and background on style selection in the forum thread https://musescore.org/en/node/76561

__Data-wise__ changes:

Adds __4 new styles of fret marks__ for French tablatures:
- "Phalèse" from mid-XVI c. Phalèse prints (replaces the "Late Renaissance" style which was unsuccessful)
- "Bonneuil-de Visée" from 1686 print by Bonneuil of de Visée "Livre de pièces pour la guittarre"
- "Bonneuil-Gaultier" from late XVII prints by bonneuil of Gaultier works
- "Lute Didactic", exampled on Yisrael's well-known methods for lute, an 'abstract' style familiar to lutenists

Adds __two new styles of duration symbols_ of French origin:
- "French Baroque" which goes well with the "Bonneuil-de Visée" fret mark style;
- "French Baroque (headless)" which goes well with the "Bonneuil-de Visée" style.

__Code-wise__ changes:

Converts legacy names into newer ones.

3. TAB - Font design improvement

- "Lute Didactic" style scaled to be similar to other styles.
- Improved vertical alignments of glyph within each style.
- Several small fine tuning of individual shapes.

Only the font is changed; no changes to code.

4. TAB: New historical English fret style

Reference: https://musescore.org/en/node/82061 (forum thread)

Upon user request, adds a new character style for fret marks taken from a type used in several prints by different publishers in London between the last years of 16th c. and the beginning of 17th c.

Among them:
- Corkine's two lute books
- Some works by J. Dowland
- R. Dowland's _Varietie of Lute Lessons_
- Pilkington's _Ayres_
- Robinson's _Schoole of Musicke_

### CODE REVIEW

**Comment by MarkusLutz63** [comment]
... I'm not sure: Is this PR including the one, that I made before? I build it on a copy of that branch ...

File: libmscore/stafftype.h [hunk 1]
```diff
@@ -282,14 +320,15 @@ class StaffType {
       qreal durationFontSize() const      { return _durationFontSize;   }
       qreal durationFontUserY() const     { return _durationFontUserY;  }
       qreal durationFontYOffset()         { setDurationMetrics(); return _durationYOffset + _durationFontUserY * SPATIUM20; }
+      qreal durationGridYOffset()         { setDurationMetrics(); return _durationGridYOffset;}
       qreal fretBoxH()                    { setFretMetrics(); return _fretBoxH; }
-      qreal fretBoxY()                    { setFretMetrics(); return _fretBoxY + _fretFontUserY * SPATIUM20; }
+      qreal fretBoxY()                    { setFretMetrics(); return _fretBoxY + _fretFontUserY * DPI * SPATIUM20; }
```
Comment (MarkusLutz63) [inline libmscore/stafftype.h]: Probably I have to delete DPI here (don't know, why I left it), as there is a problem with fretFontUserY!

File: mscore/scoreview.cpp [hunk 2]
```diff
@@ -1458,7 +1458,7 @@ void ScoreView::moveCursor()
             }
       double x        = segment->canvasPos().x();
       double y        = system->staffYpage(staffIdx) + system->page()->pos().y();
-      Staff* staff    = _score->staff(staffIdx);
+      Staff* staff            = _score->staff(staffIdx);
```
Comment (MarkusLutz63) [inline mscore/scoreview.cpp]: This change also has to be reverted ...

**Comment by Jojo-Schmitz** [comment]
This PR indeed includes #3063, so probably needs to get debased once that got merged.

File: libmscore/chord.cpp [hunk 3]
```diff
@@ -1474,12 +1481,12 @@ void Chord::layoutStem()
...
-                        return;
-                        }
-                  // if stems are through staff, use standard formatting
+                         }
+                  return;
                   }
+            // if stems are through staff, use standard formatting
             }
-
+
```
Comment (Jojo-Schmitz) [inline libmscore/chord.cpp]: Whitespace damage

File: libmscore/stafftype.h [hunk 4]
```diff
@@ -112,9 +136,14 @@ struct TablatureDurationFont {
...
       qreal   defYOffset;             // the default Y displacement
+      qreal   gridBeamWidth  = GRID_BEAM_DEF_WIDTH;   // the width of the 'grid'-style beam (in sp)
+      qreal   gridStemHeight = GRID_STEM_DEF_HEIGHT;  // the height of the 'grid'-style stem (in sp)
+      qreal   gridStemWidth  = GRID_STEM_DEF_WIDTH;   // the width of the 'grid'-style stem (in sp)
+      // the note value with no beaming in 'grid'-style beaming
+      TDuration::DurationType zeroBeamLevel = TDuration::DurationType::V_QUARTER;
       QChar   displayDot;             // the char to use to draw a dot
       QChar   displayValue[int(TabVal::NUM_OF)];       // the char to use to draw a duration value
-
+
```
Comment (Jojo-Schmitz) [inline libmscore/stafftype.h]: Whitespace damage

File: mscore/scoreview.cpp [hunk 5]
```diff
@@ -3309,6 +3315,7 @@ void ScoreView::cmd(const QAction* a)
             cmdAddFret(13);
       else if(cmd == "fret-14")
             cmdAddFret(14);
+
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: Better without that newline

File: libmscore/stafftype.cpp [hunk 6]
```diff
@@ -707,15 +857,19 @@ TabDurationSymbol::TabDurationSymbol(Score* s)
...
+      _beamLength = 0.0;
+      _tab        = 0;
+      _text       = QString();
       }
-
-TabDurationSymbol::TabDurationSymbol(Score* s, StaffType* tab, TDuration::DurationType type, int dots)
-   : Element(s)
+ 
+ TabDurationSymbol::TabDurationSymbol(Score* s, StaffType* tab, TDuration::DurationType type, int dots)
+    : Element(s)
```
Comment (Jojo-Schmitz) [inline libmscore/stafftype.cpp]: Whitespace damage

File: mscore/scoreview.cpp (hunk 5 above)
Comment (MarkusLutz63) [inline mscore/scoreview.cpp]: No problem, I can delete it - thank you for your comments!

**Comment by MarkusLutz63** [comment]
These 2 PR will include the tablature code for most lute instruments, that have more than 6 strings.
Most of it works quite fine. As I pointed out above, there is a problem with user fretFontUserY, which probably is fixed easily by deleting the DPI. 
But there is another bug, that probably even isn't from that commits: If the TAB has different voices, it seems that currently the code can corrupt pieces. I have seen, that there had been bug fixes regarding voice problems. But I don't know know, if that will fix that problem,  which is quite severe, as corrupted scores are quite a problem ... 
Possibly I need some help there - but first, after correcting the discussed places, I will again search through the commited fixes. Maybe there is already a solution

Further there are some things, that could be improved, that doesn't have to do directly with the code I committed: 
- size of time signs (it would be good to also have a solution in 2.1; the 3.0-solution is excellent)
- size and placement of ornaments
- improvement of placement of bourdons, if the tab signs are on the lines etc.
- also I miss really bad a pointed line (points like in the repeat bar lines, but in every distance between the lines: 4 for notation, 5 for tabs), that is used directly after or before  bar lines or within a bar to mark petit reprises (so it should work as 1. an attachment to barlines and 2. as a sort of barline, if used within a measure). That line was quite common in baroque times in tablature, but also in notation (and it is nicer to see exactly, where a reprise begins)

**Comment by MarkusLutz63** [comment]
One remark from my side: It seems as if it works quite fine in Version 2.1 so far! 

There are some problems, if I try to open files, that I saved with version 3: 
1. Sometimes I changed singular bass notes to voice 2 to be able to adjust slurs as needed (normally they are below the signs in lute notation), if I do that in version 2.1 all is good. If I do the same in Version 3, that will give a mismatch in 2.1  (the notes of voice 2 will be appended to the measure ...).
2. There seem to be gaps in some music, that I don't know how to delete, if I open scores from 3  in 2.1 

But as these are problems that are not really relevant (scores of version 3 are not to be compatible with 2.1, isn't it), I think the code could be merged - at least from my side ;-).

**Comment by lasconic** [comment]
*scores of version 3 are not to be compatible with 2.1, isn't it)*
Version 3 doesn't exist :)

The actual issue is Scores from 2.0.3 and scores made in 2.1 in MuseScore 2.0.3.

**Comment by MarkusLutz63** [comment]
OK :)  - I meant the development version.
 
As far as I can see, there are no issues opening scores made with 2.0.3 in version 2.1.

The other way round, the bass strings won't be seen in 2.0.3, and another font is used, but the basses even are saved within 2.0.3., if I open a score from 2.1 in 2.0.3. 
If I open it again in 2.1, the bass notes are slightly shifted and a wrong font is used, but I haven't seen a crash or something like that. 

For sure it's necessary to document that differences, but I don't think that anyone, who wants to write baroque lute or archiliuto scores will use 2.0.3, if 2.1 is available. And it is only a lute thing ...

**Comment by lasconic** [comment]
This is merged.

**Pull request closed without merging**